src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1198 @@
+#include "VST3OptionsDialog.h"
+
+#ifdef __WXMSW__
+#include <ShlObj_core.h>
+#elif __WXGTK__
+#include "internal/x11/SocketWindow.h"
+#endif
+
+namespace {
+
+Steinberg::Vst::IHostApplication& LocalContext()
+{
+   static Steinberg::Vst::HostApplication localContext;
+   return localContext;
+}
+
+class PresetsBufferStream : public Steinberg::Vst::BufferStream
+{
+public:
+
+   static Steinberg::IPtr<PresetsBufferStream> fromString(const wxString& str)
+   {
+      Steinberg::Buffer buffer(str.size() / 4 * 3);
+      auto len = Base64::Decode(str, buffer);
+      wxASSERT(len <= buffer.getSize());
+      buffer.setSize(len);
+
+      auto result = owned(safenew PresetsBufferStream);
+      result->mBuffer.take(buffer);
+      return result;
+   }
+
+   wxString toString() const
+   {
+      return Base64::Encode(mBuffer, mBuffer.getSize());
+   }
+
+};
+
+void ActivateBuses(Steinberg::Vst::IComponent* component, const Steinberg::Vst::MediaType mediaType, const Steinberg::Vst::BusDirection direction, bool state)
+{
+   using namespace Steinberg;
+
+   Vst::BusInfo busInfo;
+   for(int i = 0, count = component->getBusCount(mediaType, direction); i < count; ++i)
+      if(component->getBusInfo(mediaType, direction, i, busInfo) == kResultOk &&
+         busInfo.flags & Vst::BusInfo::kDefaultActive)
+      {
+         component->activateBus(mediaType, direction, i, state ? 1 : 0);
+      }
+}
+
+wxString GetFactoryPresetsBasePath()
+{
+#ifdef __WXMSW__
+   WCHAR commonFolderPath[MAX_PATH];
+   if(SHGetFolderPathW(NULL, CSIDL_COMMON_APPDATA, NULL, 0, commonFolderPath) == S_OK)
+   {
+      return wxString(commonFolderPath) + "\\VST3 Presets\\";
+   }
+   return {};
+#elif __WXMAC__
+   return wxString("Library/Audio/Presets/");
+#elif __WXGTK__
+   return wxString("/usr/local/share/vst3/presets/");
+#endif
+}
+
+wxString GetPresetsPath(const wxString& basePath, const VST3::Hosting::ClassInfo& effectClassInfo)
+{
+   wxRegEx fixName(R"([\\*?/:<>|])");
+   wxString companyName = wxString (effectClassInfo.vendor()).Trim();
+   wxString pluginName = wxString (effectClassInfo.name()).Trim();
+
+   fixName.ReplaceAll( &companyName, { "_" });
+   fixName.ReplaceAll( &pluginName, { "_" });
+
+   wxFileName result;
+   result.SetPath(basePath);
+   result.AppendDir(companyName);
+   result.AppendDir(pluginName);
+   auto path = result.GetPath();
+
+   return path;
+}
+
+wxString GetFactoryPresetsPath(const VST3::Hosting::ClassInfo& effectClassInfo)
+{
+   return GetPresetsPath(
+      GetFactoryPresetsBasePath(),
+      effectClassInfo
+   );
+}
+
+}
+
+void VST3Effect::Initialize()
+{
+   using namespace Steinberg;
+
+   const auto& pluginFactory = mModule->getFactory();
+   if(auto effectComponent = pluginFactory.createInstance<Vst::IComponent>(mEffectClassInfo.ID()))
+   {
+      if(effectComponent->initialize(&LocalContext()) == kResultOk)
+      {
+         if(auto audioProcessor = FUnknownPtr<Vst::IAudioProcessor>(effectComponent))
```

Comment:
The commit that introduces `FUnknownPtr` can be squashed as a fixiup of
"Plugin GUI & processing fixes"